pkg/authn/delegating.go

```diff
@@ -29,29 +28,23 @@ import (
 )
 
 // NewDelegatingAuthenticator creates an authenticator compatible with the kubelet's needs
-func NewDelegatingAuthenticator(client authenticationclient.TokenReviewInterface, authn *AuthnConfig) (authenticator.Request, error) {
+func NewDelegatingAuthenticator(ctx context.Context, client authenticationclient.TokenReviewInterface, authn *AuthnConfig) (authenticator.Request, error) {
 	if client == nil {
 		return nil, errors.New("tokenAccessReview client not provided, cannot use webhook authentication")
 	}
 
 	var (
-		p   authenticatorfactory.CAContentProvider
+		p   *dynamiccertificates.DynamicFileCAContent
 		err error
 	)
 	if len(authn.X509.ClientCAFile) > 0 {
-		if len(authn.X509.ClientCAFile) == 0 {
-			return nil, fmt.Errorf("missing filename for ca bundle")
-		}
-
-		caBundle, err := ioutil.ReadFile(authn.X509.ClientCAFile)
+		p, err = dynamiccertificates.NewDynamicCAContentFromFile("client-ca", authn.X509.ClientCAFile)
 		if err != nil {
 			return nil, err
 		}
 
-		p, err = dynamiccertificates.NewStaticCAContent(authn.X509.ClientCAFile, caBundle)
-		if err != nil {
-			return nil, err
-		}
+		// worker count is ignored
+		go p.Run(1, ctx.Done())
```

Comment:
I recommend against having New methods start gofuncs.  Instead, you can return an interface that includes both authenticator.Request and a Run and start this in the caller.

You can call `RunOnce` in this method to ensure you have a valid authenticator to begin.

Reply:
agreed, and i was having a similar thought, but didn't want to complicate things. agreed, let's return an enhanced interface and start outside the constructor :+1:

Reply:
> You can call RunOnce in this method to ensure you have a valid authenticator to begin.
I think we're fine here as the bundle is loaded in the constructor already https://github.com/kubernetes/apiserver/blob/e78fb1931d1a7682240456bf1883759720b97bb1/pkg/server/dynamiccertificates/dynamic_cafile_content.go#L85-L87